### PR TITLE
[Shop] Different variant prices on product details page

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | yes
| BC breaks?      | no
| Related tickets |
| License         | MIT

On product details page, we have always one price displayed (from the first/default/etc variant). It's obviously confusing, as we could have different price for each variant - therefore it should change after changing variant/option. In this PR, I will implement some lovely JS script, to handle this logic and make *Sylius* product details page more and more perfect, to reach the sky and more with its perfectness 🌠 🎉 

zrzut ekranu 2016-09-20 o 15 22 58
zrzut ekranu 2016-09-20 o 15 23 03
zrzut ekranu 2016-09-20 o 15 23 26
zrzut ekranu 2016-09-20 o 15 24 09

### CODE REVIEW

File: features/product/viewing_products/viewing_different_price_for_different_product_variants.feature [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+Feature: Viewing different price for different product variants
+    In order to see product variant price
+    As a Visitor
+    I want to be able to see proper price for each product variant
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a "Wyborowa Vodka" configurable product
+        And the product "Wyborowa Vodka" has "Wyborowa Vodka Exquisite" variant priced at "$40.00"
+        And the product "Wyborowa Vodka" has "Wyborowa Apple" variant priced at "$4.00"
```
Comment (pamil) [inline features/product/viewing_products/viewing_different_price_for_different_product_variants.feature]: Their prices differ by just one zero, can we use some other numbers than 4 and 0 to make the difference more clear?

File: src/Sylius/Bundle/CoreBundle/Controller/ProductController.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function showAction(Request $request)
+    {
+        $configuration = $this->requestConfigurationFactory->create($this->metadata, $request);
+
+        $this->isGrantedOr403($configuration, ResourceActions::SHOW);
+        /** @var ProductInterface $resource */
+        $resource = $this->findOr404($configuration);
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Controller/ProductController.php]: If we typehinting $resource anyway, can we change a name of variable?

File: src/Sylius/Bundle/CoreBundle/Controller/ProductController.php [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    /**
+     * @param ProductInterface $resource
+     * @param RequestConfiguration $configuration
+     * @param View $view
+     *
+     * @return View
+     */
+    protected function prepareHtmlRequestView(ProductInterface $resource, RequestConfiguration $configuration, View $view)
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Controller/ProductController.php]: `$resource` -> `$product`

File: src/Sylius/Bundle/CoreBundle/Controller/ProductController.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        $templateData = [
+            'configuration' => $configuration,
+            'metadata' => $this->metadata,
+            'resource' => $resource,
+            $this->metadata->getName() => $resource,
+        ];
+
+        if (
+            !$resource->isSimple() &&
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Controller/ProductController.php]: Creating `$product->isConfigurable()` method which is a shortcut for `!$product->isSimple()` would improve readability a lot.

File: src/Sylius/Bundle/ShopBundle/Resources/private/js/variants-prices.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        handleProductVariantsChange();
+    }
+});
+
+function handleProductOptionsChange() {
+    $('[name*="sylius_cart_item[variant]"]').on('change', function() {
+        $selector = '';
+
+        $('#sylius-product-adding-to-cart select').each(function() {
+            $selector += '[data-'+$(this).attr('data-option')+'="'+$(this).find('option:selected').text()+'"]';
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ShopBundle/Resources/private/js/variants-prices.js]: Missing spaces before and after each '+'

File: src/Sylius/Component/Core/Provider/ProductVariantsPricesProvider.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+namespace Sylius\Component\Core\Provider;
+
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Sylius\Component\Product\Model\OptionValueInterface;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class ProductVariantsPricesProvider implements ProductVariantsPricesProviderInterface
```
Comment (Arminek) [inline src/Sylius/Component/Core/Provider/ProductVariantsPricesProvider.php]: final?

File: src/Sylius/Behat/Context/Setup/ProductContext.php [hunk 7]
```diff
@@ -153,6 +153,11 @@ public function storeHasAConfigurableProduct($productName)
         $product->setCode($this->convertToCode($productName));
         $product->setDescription('Awesome '.$productName);
 
+        if ($this->sharedStorage->has('channel')) {
+            $channel = $this->sharedStorage->get('channel');
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/ProductContext.php]: `$channel` is not used anywhere else. ` $product->addChannel($this->sharedStorage->get('channel'));`

File: features/product/viewing_products/viewing_different_price_for_different_product_variants.feature [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a "Wyborowa Vodka" configurable product
+        And the product "Wyborowa Vodka" has "Wyborowa Vodka Exquisite" variant priced at "$40.00"
+        And the product "Wyborowa Vodka" has "Wyborowa Apple" variant priced at "$12.55"
+
+    @ui
+    Scenario: Viewing a detailed page with default variant's price
+        When I check product "Wyborowa Vodka" details
```
Comment (michalmarcinkowski) [inline features/product/viewing_products/viewing_different_price_for_different_product_variants.feature]: When I view product "Wyborowa Vodka"

File: features/product/viewing_products/viewing_different_price_for_different_product_variants.feature [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        Then I should see the product price "$40.00"
+
+    @ui @javascript
+    Scenario: Viewing a detailed page with product's price for different variant
+        When I check product "Wyborowa Vodka" details
+        And I select "Wyborowa Apple" variant
+        Then I should see the product price "$12.55"
```
Comment (michalmarcinkowski) [inline features/product/viewing_products/viewing_different_price_for_different_product_variants.feature]: Then the product price should be "$12.55"

**Comment by michalmarcinkowski** [comment]
We need one more Behat scenario:
```
Give the store has a "Wyborowa Vodka" configurable product
And this product has option "Volume" with values "0.5L", "0.7L" and "1L"
And this product is available in "0.5L" volume priced at "$20.00"
And this product is available in "0.7L" volume priced at "$25.00"
When I view this product
And I select "1L" option
Then the product price should be "unavailable"
```

But this can be implemented in a separate PR.

**Comment by michalmarcinkowski** [comment]
@Zales0123 one last fix required to make it green.

**Comment by Zales0123** [comment]
@michalmarcinkowski ready to go ;)

File: src/Sylius/Component/Core/Provider/ProductVariantsPricesProvider.php [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+     *
+     * @return array
+     */
+    private function constructOptionsMap(ProductVariantInterface $variant)
+    {
+        $optionMap = [];
+
+        /** @var OptionValueInterface $option */
+        foreach ($variant->getOptions() as $option) {
+            $optionMap[$option->getOption()->getCode()] = $option->getValue();
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Provider/ProductVariantsPricesProvider.php]: We can use `$option->getOptionCode()`

**Comment by Zales0123** [comment]
@pjedrzejewski seems reasonable for me 👍

File: src/Sylius/Bundle/CoreBundle/Twig/ProductVariantsPricesExtension.php [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        $this->productVariantsPricesHelper = $productVariantsPricesHelper;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFunctions()
+    {
+        return [
+            new \Twig_SimpleFunction('sylius_product_variant_prices', [$this, 'getVariantsPrices']),
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Twig/ProductVariantsPricesExtension.php]: Won't it be simpler to use `productVariantsPricesHelper` directly? 
`new \Twig_SimpleFunction('sylius_product_variant_prices', [$this->productVariantsPricesHelper, 'getPrices'])`

**Comment by pjedrzejewski** [comment]
Thanks Mateusz! :) Nice work!

**Comment by Zales0123** [comment]
👍  I will apply last comments in additional PR in a minute.

**Pull request merged**